### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by sharad-epam**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+node_modules
+
+# testing
+/coverage
+
+**/__snapshots__/**
+
+# build files
+/build
+.vscode/
```
Comment (GlebShaucou) [inline .gitignore]: add dist folder to gitignore

File: src/js/components/presentational/Search.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    const { searchBySelected } = this.props;
+    return (
+      <div>
+        <input
+          placeholder="Search Movies..."
+          name="search"
+          className="search-input"
+          onChange={this.onChange}
+        />
+        <button className="btnSearch" onClick={() => this.selectedVal()}>
```
Comment (GlebShaucou) [inline src/js/components/presentational/Search.js]: button could be a separate component

File: src/js/components/presentational/Search.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    this.props.updatedSearchBySelected("Title");
+  };
+  handleGenreClick = () => {
+    this.props.updatedSearchBySelected("Genre");
+  };
+  render() {
+    const { searchBySelected } = this.props;
+    return (
+      <div>
+        <input
```
Comment (GlebShaucou) [inline src/js/components/presentational/Search.js]: input could be a separate component

File: src/js/components/presentational/Search.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import React, { Component } from "react";
```
Comment (GlebShaucou) [inline src/js/components/presentational/Search.js]: use PropTypes to make code more readable and components more reusable.

File: src/js/components/presentational/Movies.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            <span className="genre"> {item.release_date}</span>
+            <span className="genre">Vote Avg:({item.vote_average})</span>
+          </div>
+        </div>
+      );
+    });
+  };
+  render() {
+    if (this.state.error) {
+      throw new Error("No movies found");
```
Comment (GlebShaucou) [inline src/js/components/presentational/Movies.js]: render method of react components should return JSX, throw is a very bad practice.

File: src/js/components/presentational/Movies.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    if (this.props.movies.length === 0) {
+      this.setState({
+        error: true
+      });
+    }
+  }
+  renderMovies = () => {
+    return this.props.movies.map(item => {
+      return (
+        <div className="movieList" key={item.id}>
```
Comment (GlebShaucou) [inline src/js/components/presentational/Movies.js]: single movie representation could be a separate component.

File: src/js/components/presentational/Movies.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import React, { Component } from "react";
+import "../../styles/App.css";
+
+class Movies extends Component {
+  state = {
+    error: false
+  };
+  componentDidMount() {
+    if (this.props.movies.length === 0) {
```
Comment (GlebShaucou) [inline src/js/components/presentational/Movies.js]: it is not an error if no movies found. show instead simple message that no movies found.

File: src/js/components/presentational/Header.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import React from "react";
+import "../../styles/Header.css";
+
+const Header = (props) => {
+  const { moviesFound, sortBySelected, handleReleaseDateClick, handleRatingClick } = props;
+  return (
+    <div className="headerWrapper">
+      <div className="moviesFound">{moviesFound} movies found</div>
+      <div className="sortBy"><span>SORT BY</span>
+      <button className={`${sortBySelected === 'ReleaseDate' ? ' btn-selected' : ''}`} onClick={handleReleaseDateClick}>RELEASE DATE</button>
```
Comment (GlebShaucou) [inline src/js/components/presentational/Header.js]: button could be a separate component.

File: src/js/components/container/App.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    return filteredMovieListSortBy;
+  };
+  render() {
+    const { searchBySelected, sortBySelected } = this.state;
+    const filteredMovies = this.filterMovies() || [];
+    const moviesFound = filteredMovies.length;
+    return (
+      <div className="container">
+        <ErrorBoundry>
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: this code looks to heavy. there are two possible solutions:
1. you can use single top level ErrorBoundry;
2. you can create HOC (https://reactjs.org/docs/higher-order-components.html) that will wrap your components with ErrorBoundry, this approach will make your code more flexible.

File: src/js/components/container/App.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  updatedSearchBySelected = searchBy => {
+    this.setState({ searchBySelected: searchBy });
+  };
+  handleReleaseDateClick = () => {
+    this.setState({ sortBySelected: "ReleaseDate" });
+  };
+  handleRatingClick = () => {
+    this.setState({ sortBySelected: "Rating" });
+  };
+  filterMovies = () => {
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: create a separate pure utils function that will filter your movies.

File: src/js/components/container/App.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+import SearchBar from "../presentational/Search";
+import myData from "../../../../movies.json";
+import Header from "../presentational/Header";
+import Movies from "../presentational/Movies";
+import ErrorBoundry from "../presentational/ErrorBoundry";
+import "../../styles/App.css";
+
+class App extends Component {
+  constructor() {
+    super();
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: where are props?

File: src/js/components/container/App.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
+import React, { Component } from "react";
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: use PropTypes to make your components reusable.

File: src/js/components/container/App.js [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    super();
+    this.state = {
+      filterText: "",
+      searchBySelected: "Title", // Title, Genre
+      sortBySelected: "ReleaseDate" // ReleaseDate, Rating
+    };
+  }
+  handleUserInput = filterText => {
+    this.setState({
+      filterText: filterText
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: use es6 features here.

File: src/js/components/container/App.js [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    this.setState({ sortBySelected: "ReleaseDate" });
+  };
+  handleRatingClick = () => {
+    this.setState({ sortBySelected: "Rating" });
+  };
+  filterMovies = () => {
+    const { filterText, searchBySelected, sortBySelected } = this.state;
+    let filteredMovieList = [];
+    let filteredMovieListSortBy = [];
+    if (filterText) {
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: make 3 different pure functions (https://medium.com/javascript-scene/master-the-javascript-interview-what-is-a-pure-function-d1c076bec976). each function should fulfill their own work.

File: src/js/components/container/App.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+          <Movies movies={filteredMovies} />
+        </ErrorBoundry>
+      </div>
+    );
+  }
+}
+export default App;
+
+const wrapper = document.getElementById("create-article-form");
+wrapper ? ReactDOM.render(<App />, wrapper) : false;
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: what is the purpose of `false`? why do you need such expression?

File: src/js/components/presentational/ErrorBoundry.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
+import React from "react";
+class ErrorBoundary extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { hasError: false };
+  }
+  static getDerivedStateFromError(error) {
```
Comment (GlebShaucou) [inline src/js/components/presentational/ErrorBoundry.js]: what does this method do?

File: src/js/components/presentational/ErrorBoundry.js [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import React from "react";
```
Comment (GlebShaucou) [inline src/js/components/presentational/ErrorBoundry.js]: you have a different code style along all application. in some places you use Component in some other places React.Component. you should be more consistently.

File: src/js/components/presentational/Header.js [hunk 18]
```diff
@@ -0,0 +1,15 @@
+import React from "react";
```
Comment (GlebShaucou) [inline src/js/components/presentational/Header.js]: use PropTypes.

File: src/js/components/presentational/Movies.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
+import React, { Component } from "react";
```
Comment (GlebShaucou) [inline src/js/components/presentational/Movies.js]: use PropTypes.

File: src/js/components/presentational/Movies.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+import "../../styles/App.css";
```
Comment (GlebShaucou) [inline src/js/components/presentational/Movies.js]: create a separate styles file for Movies ex. Movies.css

File: src/js/components/container/App.js [hunk 21]
```diff
@@ -0,0 +1,106 @@
+import React, { Component } from "react";
+import ReactDOM from "react-dom";
+import SearchBar from "../presentational/Search";
+import myData from "../../../../movies.json";
+import Header from "../presentational/Header";
+import Movies from "../presentational/Movies";
+import ErrorBoundry from "../presentational/ErrorBoundry";
+import "../../styles/App.css";
```
Comment (GlebShaucou) [inline src/js/components/container/App.js]: it is much more convenient to have styles for component in the same folder.

File: src/index.js [hunk 22]
```diff
@@ -0,0 +1 @@
+import App from "./js/components/container/App";
```
Comment (GlebShaucou) [inline src/index.js]: use App her 
`ReactDOM.render(<App />, document.getElementById("create-article-form"))`

File: src/js/components/presentational/Search.js [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+import "../../styles/Search.css";
+class SearchBar extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      search: ""
+    };
+  }
+  selectedVal = e => {
```
Comment (GlebShaucou) [inline src/js/components/presentational/Search.js]: were e is used?

File: webpack.config.js [hunk 24]
```diff
@@ -0,0 +1,32 @@
+const HtmlWebPackPlugin = require("html-webpack-plugin");
+module.exports = {
```
Comment (GlebShaucou) [inline webpack.config.js]: configure entry and output https://webpack.js.org/guides/getting-started/

File: .gitignore (hunk 1 above)
Comment (sharad-epam) [reply .gitignore]: Added

File: src/js/components/presentational/Search.js (hunk 2 above)
Comment (sharad-epam) [reply src/js/components/presentational/Search.js]: Separated

File: src/js/components/presentational/Search.js (hunk 3 above)
Comment (sharad-epam) [reply src/js/components/presentational/Search.js]: Separated

File: src/js/components/presentational/Search.js (hunk 4 above)
Comment (sharad-epam) [reply src/js/components/presentational/Search.js]: done

File: src/js/components/presentational/Movies.js (hunk 5 above)
Comment (sharad-epam) [reply src/js/components/presentational/Movies.js]: removed

File: src/js/components/presentational/Movies.js (hunk 6 above)
Comment (sharad-epam) [reply src/js/components/presentational/Movies.js]: Separated a component

File: src/js/components/presentational/Movies.js (hunk 7 above)
Comment (sharad-epam) [reply src/js/components/presentational/Movies.js]: Yes modified

File: src/js/components/presentational/Header.js (hunk 8 above)
Comment (sharad-epam) [reply src/js/components/presentational/Header.js]: separated

File: src/js/components/container/App.js (hunk 9 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: Yes used a single ErrorBoundry on Top Level

File: src/js/components/container/App.js (hunk 10 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: separated and made pure functions

File: src/js/components/container/App.js (hunk 11 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: got that modified

File: src/js/components/container/App.js (hunk 12 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: yes used

File: src/js/components/container/App.js (hunk 13 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: yes used

File: src/js/components/container/App.js (hunk 14 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: separated the functions into pure functions

File: src/js/components/container/App.js (hunk 15 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: removed

File: src/js/components/presentational/ErrorBoundry.js (hunk 16 above)
Comment (sharad-epam) [reply src/js/components/presentational/ErrorBoundry.js]: This methods sets tje hasError value true and the below method
  componentDidCatch(error, info) {
    // You can also log the error to an error reporting service
   // logErrorToMyService(error, info);
  }
will log the errors if hasError true
wrote as per the doc

File: src/js/components/presentational/ErrorBoundry.js (hunk 17 above)
Comment (sharad-epam) [reply src/js/components/presentational/ErrorBoundry.js]: Yes that was a mistake , corrected

File: src/js/components/presentational/Header.js (hunk 18 above)
Comment (sharad-epam) [reply src/js/components/presentational/Header.js]: used

File: src/js/components/presentational/Movies.js (hunk 19 above)
Comment (sharad-epam) [reply src/js/components/presentational/Movies.js]: used

File: src/js/components/presentational/Movies.js (hunk 20 above)
Comment (sharad-epam) [reply src/js/components/presentational/Movies.js]: yes separated

File: src/js/components/container/App.js (hunk 21 above)
Comment (sharad-epam) [reply src/js/components/container/App.js]: yeah following that convention

File: src/index.js (hunk 22 above)
Comment (sharad-epam) [reply src/index.js]: yes used

File: src/js/components/presentational/Search.js (hunk 23 above)
Comment (sharad-epam) [reply src/js/components/presentational/Search.js]: removed since we are not using

File: webpack.config.js (hunk 24 above)
Comment (sharad-epam) [reply webpack.config.js]: yes added

File: src/js/components/container/MoviesContainer.js [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+import ReactDOM from "react-dom";
+import SearchBar from "../presentational/Search/Search";
+import myData from "../../../../movies.json";
+import Header from "../presentational/Header/Header";
+import Movies from "../presentational/Movies/Movies";
+import ErrorBoundry from "../presentational/ErrorBoundry";
+
+class MoviesContainer extends Component {
+  constructor() {
+    super();
```
Comment (GlebShaucou) [inline src/js/components/container/MoviesContainer.js]: if you do not use props inside constructor then state can be declared inside component.

File: src/js/components/presentational/Movies/Movie/Movie.js [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+        <span key={index}> {genre}</span>
+      ))}
+      <span className="genre"> {movie.release_date}</span>
+      <span className="genre">Vote Avg:({movie.vote_average})</span>
+    </div>
+  </div>
+);
+
+Movie.propTypes = {
+  movies: PropTypes.object
```
Comment (GlebShaucou) [inline src/js/components/presentational/Movies/Movie/Movie.js]: Inside your component you use movie prop. Pay more attention what you are  doing.

File: dist/main.js [hunk 27]
```diff
@@ -0,0 +1,31 @@
+!function(e){var t={};function n(r){if(t[r])return t[r].exports;var l=t[r]={i:r,l:!1,exports:{}};return e[r].call(l.exports,l,l.exports,n),l.l=!0,l.exports}n.m=e,n.c=t,n.d=function(e,t,r){n.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},n.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},n.t=function(e,t){if(1&t&&(e=n(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(n.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var l in e)n.d(r,l,function(t){return e[t]}.bind(null,l));return r},n.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return n.d(t,"a",t),t},n.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},n.p="",n(n.s=11)}([function(e,t,n){"use strict";e.exports=n(4)},function(e,t,n){e.exports=n(9)()},function(e,t,n){"use strict";
```
Comment (GlebShaucou) [inline dist/main.js]: remove dist from PR

File: dist/index.html [hunk 28]
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
```
Comment (GlebShaucou) [inline dist/index.html]: remove dist from PR

File: src/App.js [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+import { PageNotFound } from "./js/components/presentational/PageNotFound";
+class App extends Component {
+  render() {
+    return (
+      <Provider store={store}>
+        <Router history={history}>
+          <Switch>
+            <Route path="/" exact component={MoviesContainer} />
+            <Route path="/movie/:id" component={MovieDetails} />
+            <Route component={PageNotFound} />
```
Comment (GlebShaucou) [inline src/App.js]: Using direct link in a new tab to movie does not works.
Search routing does not implemented correctly.
Can not return to main page after visiting 404 page.

File: src/js/components/presentational/ErrorBoundry.js [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+class ErrorBoundary extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { hasError: false };
+  }
+  static getDerivedStateFromError(error) {
+    return { hasError: true };
+  }
+
+  componentDidCatch(error, info) {
```
Comment (GlebShaucou) [inline src/js/components/presentational/ErrorBoundry.js]: Do you need this method here?

Comment (sharad-epam) [reply src/js/components/presentational/ErrorBoundry.js]: Yeah we can keep either componentDidCatch or getDerivedStateFromError, will remove one

**Pull request merged**